### PR TITLE
functional: configure machines with cloud-config

### PR DESCRIPTION
**Pull request by marineam**
In preparation for running the fleet functional tests on other platforms
this moves most instance configuration into cloud config. This is not
quite ready to merge, the network configuration is nspawn specific and
overall runs about twice as slow as before and I'm not sure why yet.
Posting for any early feedback if anyone is interested.

### CODE REVIEW

File: functional/util/config.go [hunk 1]
```diff
@@ -0,0 +1,169 @@
+/*
```
Comment (jonboulle) [inline functional/util/config.go]: Please switch to line comments, we just did this elsewhere (block comments interfere with build tags :/)

**Comment by ecnahc515** [comment]
+1 I was just considering doing this myself the other day. What are the possibilities of adding a Vagrantfile for functional tests?

**Comment by marineam** [comment]
Updated, depends on https://github.com/coreos/fleet/pull/1125 and https://github.com/coreos/fleet/pull/1124

The reason this change makes the tests run slower is encoding a big Go binary in YAML significantly slows down coreos-cloudinit apparently.

**Comment by marineam** [comment]
I think the most practical way to leverage cloud-config but not suffer the slowdown would be to support multi-part mime as an input format: http://cloudinit.readthedocs.org/en/latest/topics/format.html#mime-multi-part-archive

But that is not a quick thing to add. Any opinion on whether the slowdown is acceptable or if the alternative of nspawn keeping the existing setup scheme and only new platforms use cloud config is preferable? The latter just introduces a larger difference in behavior between platforms.

**Comment by jonboulle** [comment]
Can we compromise for now and copy the binaries in but use cloudconfig for everything else?

**Comment by marineam** [comment]
@jonboulle updated, making the inline fleetd optional since I still want that for other things.

File: functional/platform/nspawn.go [hunk 2]
```diff
@@ -225,46 +228,41 @@ func (nc *nspawnCluster) prepCluster() (err error) {
 	return nil
 }
 
-func (nc *nspawnCluster) prepFleet(dir, ip, sshKeySrc, fleetdBinSrc string) error {
+func (nc *nspawnCluster) prepFleet(dir string) error {
```
Comment (jonboulle) [inline functional/platform/nspawn.go]: maybe `insertFleetd`

File: functional/platform/nspawn.go [hunk 3]
```diff
@@ -225,46 +228,41 @@ func (nc *nspawnCluster) prepCluster() (err error) {
...
 
-	relSSHKeyDst := path.Join("opt", "fleet", "id_rsa.pub")
-	sshKeyDst := path.Join(dir, relSSHKeyDst)
-	if err := copyFile(sshKeySrc, sshKeyDst, 0644); err != nil {
-		return err
-	}
-
 	fleetdBinDst := path.Join(dir, "opt", "fleet", "fleetd")
-	if err := copyFile(fleetdBinSrc, fleetdBinDst, 0755); err != nil {
+	if err := copyFile(fleetdBinPath, fleetdBinDst, 0755); err != nil {
```
Comment (jonboulle) [inline functional/platform/nspawn.go]: return copyFile

File: functional/util/config.go [hunk 4]
```diff
@@ -0,0 +1,161 @@
...
+	values := configValues{
+		NetworkUnit:   networkUnit,
+		PublicKey:     string(key),
+		Fleetd:        fleetd,
+		EtcdEndpoint:  etcdEndpoint,
+		EtcdKeyPrefix: etcdKeyPrefix,
+		FleetAPIPort:  fleetAPIPort,
+	}
+
+	if err = configTemplate.Execute(dst, &values); err != nil {
```
Comment (jonboulle) [inline functional/util/config.go]: return configTemplate.Execute....

**Comment by jonboulle** [comment]
few tiny things, but lgtm!

**Comment by bcwaldon** [comment]
Superseded by https://github.com/coreos/fleet/pull/1100

**Comment by bcwaldon** [comment]
er uh https://github.com/coreos/fleet/pull/1185

**Pull request closed without merging**